package/MDAnalysis/lib/c_gridsearch.pyx

```diff
@@ -0,0 +1,647 @@
+                            ###
+                            if neighborhood.beadids == NULL:
+                                abort()
+                            ###Modified
+                            if neighborhood.beaddist == NULL:
+                                abort()
+                            ###
+                # Register the cell as checked
+                already_checked[nchecked] = cell_index
+                nchecked += 1
+
+    return neighborhood
+cdef ns_neighborhood_holder *ns_core(real[:, ::1] refcoords,
+                                     real[:, ::1] neighborcoords,
+                                     ns_grid *grid,
+                                     PBCBox box,
+                                     real cutoff) nogil:
+    cdef ns_int coordid, i, j
+    cdef ns_int ncoords = refcoords.shape[0]
+    cdef ns_int ncoords_neighbors = neighborcoords.shape[0]
+    cdef real cutoff2 = cutoff * cutoff
+    cdef ns_neighborhood_holder *holder
+
+    cdef ns_int *neighbor_buf
+    cdef ns_int buf_size, ibuf
+
+    holder = create_neighborhood_holder()
+    if holder == NULL:
+        fprintf(stderr,"FATAL: Could not allocate memory for NS holder\n",
+                sizeof(ns_int) * ncoords)
+        abort()
+
+    holder.size = ncoords
+    holder.neighborhoods = <ns_neighborhood **> malloc(sizeof(ns_neighborhood *) * ncoords)
+    if holder.neighborhoods == NULL:
+        fprintf(stderr,"FATAL: Could not allocate memory for NS holder.neighborhoods (requested: %i bytes)\n",
+                sizeof(ns_neighborhood) * ncoords)
+        abort()
+
+    # Here starts the real core and the iteration over coordinates
+    for coordid in range(ncoords):
+        holder.neighborhoods[coordid] = retrieve_neighborhood(&refcoords[coordid, XX],
+                                                              neighborcoords,
+                                                              grid,
+                                                              box,
+                                                              cutoff2)
+        holder.neighborhoods[coordid].cutoff = cutoff
+
+    return holder
+
+# Python interface
+cdef class FastNS(object):
+    cdef PBCBox box
+    cdef readonly int nthreads
+    cdef readonly real[:, ::1] coords
+    cdef real[:, ::1] coords_bbox
+    cdef readonly real cutoff
+    cdef bint prepared
+    cdef ns_grid *grid
+
+
+    def __init__(self, box):
+        if box.shape != (3, 3):
+            raise ValueError("Box must be provided as triclinic_dimensions (a 3x3 numpy.ndarray of unit cell vectors")
+
+        self.box = PBCBox(box)
+
+        self.nthreads = 1
+
+        self.coords = None
+        self.coords_bbox = None
+
+        self.cutoff = -1
+
+        self.prepared = False
+
+        self.grid = <ns_grid *> malloc(sizeof(ns_grid))
+
+
+    def __dealloc__(self):
+        #destroy_nsgrid(self.grid)

```

Comment:
From my play with @seb-buch module in the benchmark, having this line commented leads to a memory leak. When an instance of `FastNS` is destroyed by the garbage collector, `self.grid`—that is allocated directly with malloc and therefore not handled by python—is not freed. As a result, running a neighbor search on multiple frames result in all the memory being eating up. I had to hard reboot a coule of time because of that specific line being commented out.

It is likely commented for a reason, though. Maybe @seb-buch has some insight to give us on why he commented the line.

It also means we have to be careful with memory management with that PR.

Reply:
Yes I agree.  Even raising the exceptions/ abort will lead to memory leak. While I have never coded in Cython, I have couple of possible ideas based on what people have done earlier and suggested elsewhere. 
First is we can use atexit() functionality, which makes all the malloc pointers global and whenever any instance is destroyed, all the global pointers dereference themselves (at the atexit() function) before the destruction. Additionally, we can put the same function before raising/catching any exception, so all the pointers which are populated globally are de-referenced first, followed by a gracefull exit.
Another approach to catch the exceptions is using "[cysignals/signals.pxi](http://lira.no-ip.org:8080/doc/python-cysignals-doc/html/)", which catches the exceptions inside the pure C code and can be used along with `cdef [datatype] function() except *`. This can probably be coupled with atexit() as well.

Reply:
I think the line just has to be uncommented. Also, the initialization should probably go in `__cinit__` rather than `__init__`.